nixos/services/backup/syncoid: add per-command user option

NiklasGollenstede opened the pull request:
Allow specifying a different user for each `config.services.syncoid.commands`.

With privilege delegation (i.e. with a user other than `root`), and when multiple commands involve the same dataset, it is important to use a different local user for each of those commands, as otherwise permissions can get removed when one command finishes, but the others are still in progress.

This was one of the things I needed to do to make my backups reliable (the others are quite setup-specific).
I have a setup where all hosts push encrypted snapshots to one server, which then forwards them to two other servers. Especially when one of those was down, that one failing (and the service exiting, removing the permissions) kept making the other one fail as well.

I guess this is a regression of ecd32b8104e6cca16fe1b2cfb89f39a8c7c01731 (not blaming). @etu, thoughts?

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

NiklasGollenstede committed:
nixos/services/backup/syncoid: add per-command user option

Allow specifying a different user for each `config.services.syncoid.commands`.

etu reviewed (commented):
My initial impression without trying it is that it looks like a good change that is worth to implement.

I just want to try it before I give an approval :slightly_smiling_face: 

In the meanwhile before I have time to try it out I would like to ask/request for a test case for this usage since we have tests for syncoid/sanoid :slightly_smiling_face:

NiklasGollenstede committed:
nixos/syncoid: improve documentation

Add a note on SSH key ownership when using different users for each `config.services.syncoid.commands`.

NiklasGollenstede commented:
Thanks!

> I would like to ask/request for a test case for this usage

I would be willing and should be able to do that, I'm just not sure what to test here. Since this is about concurrent sends, it is inherently racy.

Doing a push from the `source` to the existing `target`, and "at the same time" also doing a second push, either just to a different pool/dataset or to a second target, would be easy to test, but I am not sure what the takeaway from that would be. With the patch (and different send users) it should always work. Without the test, it probably would have worked as well:
(I would expect that) ZFS only checks the permissions when starting a send (and maybe at each incremental snapshot in between?). So if both sends start "at the same time", before either finished, then the operation succeeds, even if they don't finish at the same time (and one of them thus removes the permissions for the other).

Building more complex test cases, maybe with syncoid-incremental sends should be possible, but I am not sure they would show reliable behavior.

Thoughts?

lopsided98 commented:
I think using `DynamicUser` as implemented in #236729 might be a better solution.

NiklasGollenstede commented:
Mostly for the general security and system administration principle of least privilege. Ideally, we would never run any service as root without restrictions, but instead run every service with the minimum set of privileges required to execute it.

In case of syncoid, we do that using an unprivileged (normal) user with just enough ZFS permissions delegated on exactly the datasets it needs.

Syncoid specifically does a whole lot of shell command crafting. I would not at all be confident that maliciously named datasets or snapshots (e.g., created by non-root users as children of a dataset they got delegated permissions in) can't make syncoid run unintended ZFS commands, or even arbitrary shell commands. We do neither want that to run as root, nor with ZFS permissions do do stuff the intended command does not require.

---

Ah. I think I misunderstood your question. You are asking why we remove the ZFS permissions after the service finishes.

We don't want every syncoid command to also have the permissions of every other syncoid command (for the reasons stated above).
I suppose that with a dedicated, per-command user, there is fairly litte point in removing the permissions, other than "it's cleaner". Our intention is not to delegate to that user, but to delegate to the service/command, so the implementation should reflect that.

NiklasGollenstede commented:
Hm. Honestly, there is no guarantee that, now or in the future, ZFS won't process concurrent sends/receives in the order they were issued in (and not actually concurrently). Also, I think the permissions are only checked when it starts the send/recv of a dataset (or snapshot), not during the transfer of the data.

---

I agree with @lopsided98, though. https://github.com/NixOS/nixpkgs/pull/236729 implements a more elegant approach to solve this issue. That PR unfortunately also still requires quite some work, esp. because it includes some unrelated changes.

Maybe because this PR is quite minimal, and we have confirmation (from @fastfailure) that the problem is affecting not just me, this get's merged for now, still with the longer therm goal of merging #236729?
(I am currently locally applying this patch to `nixos-23.11`/`-unstable`).

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/syncoid-questions/26411/3
